Firestore/Source/Local/FSTLocalStore.mm

```diff
@@ -293,29 +251,22 @@ - (FSTMaybeDocumentDictionary *)applyRemoteEvent:(FSTRemoteEvent *)remoteEvent {
       // First make sure that all references are deleted.
       if ([mapping isKindOfClass:[FSTResetMapping class]]) {
         FSTResetMapping *reset = (FSTResetMapping *)mapping;
-        [queryCache removeMatchingKeysForTargetID:targetID group:group];
-        [queryCache addMatchingKeys:reset.documents forTargetID:targetID group:group];
+        [self.dataCache resetQuery:queryData documents:reset.documents group:group];
```

Comment:
This arrangement does more work than we need to do and potentially breaks us in the future.

Every remote event can potentially contain multiple changes. Consider this initial state:

document cache contains doc1
query cache contains
  * target1 matches doc1
  * target2 is empty

Now let's say we get a remote event that contains multiple changes
  * change1 is a reset mapping, saying target1 now matches doc2 (not doc1)
  * change2 is a update of target2 to say it now matches doc1

While processing change1 you'll see doc1 become unreferenced and eager GC will delete it. While processing change2 you'll re-add it. If the document hadn't actually changed then we'll have done a whole lot of work for nothing.

How this potentially breaks us in the future is that the server wants to be able to move us from snapshot to snapshot and can assume it knows the contents of our cache as it does it. If the server assumes we have doc1 and sends change2 as a patch saying nothing changed we'll have corrupted the cache.